otherlibs/dynlink/byte/dynlink.ml

```diff
@@ -95,46 +95,54 @@ module Bytecode = struct
 
   let run_shared_startup _ = ()
 
-  let run (ic, file_name, file_digest) ~unit_header ~priv =
+  let with_lock lock f =
+    Mutex.lock lock;
+    Fun.protect f
+      ~finally:(fun () -> Mutex.unlock lock)
```

Comment:
Remark for later: I'd like to have this in `Mutex`! In the meantime, this could go in `Misc`.